### PR TITLE
Caching util methods

### PR DESCRIPTION
**Pull request by buehner**
These methods may be helpful if you are using a GeoWebCache (integrated in a GeoServer).

I'll try to add some tests tomorrow.

### CODE REVIEW

File: src/util/Caching.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+            return gwcConfig;
+        },
+
+        /**
+         * Computes the tile origin (top left) for a given extent
+         * (lower left/upper right bbox), maxResolution and tileSize that
+         * can be set as the origin on a ol.tilegrid.TileGrid instance to
+         * match a GWC gridset configured in a GeoServer. GetMap requests
+         * should use the parameter tiled=true to make use of the cache.
```
Comment (marcjansen) [inline src/util/Caching.js]: Can you add a not that this is especially important as OL want us to provide the tile origin on the upper left side? And also please hint that the resulting grid should be usable for WMTS and WMS sources (having a GWC-integration, e.g. with `TILED=true`)

File: src/util/Caching.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+         *
+         * @param {Number} tileSize The tile size in pixels
+         * @param {Number} maxResolution The max resolution
+         * @param {Array} extent The extent to use for the gridSet
+         * @return {Array} The array with two elements representing the
+         *     tileOrigin coordinate.
+         */
+        getTileOrigin: function(tileSize, maxResolution, extent) {
+            if (Ext.isEmpty(tileSize) || Ext.isEmpty(maxResolution) ||
```
Comment (marcjansen) [inline src/util/Caching.js]: The tilesize should fallback to `[256, 256]`

File: src/util/Caching.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        },
+
+        /**
+         * Computes the tile origin (top left) for a given extent
+         * (lower left/upper right bbox), maxResolution and tileSize that
+         * can be set as the origin on a ol.tilegrid.TileGrid instance to
+         * match a GWC gridset configured in a GeoServer. GetMap requests
+         * should use the parameter tiled=true to make use of the cache.
+         *
+         * @param {Number} tileSize The tile size in pixels
```
Comment (marcjansen) [inline src/util/Caching.js]: One does not need to have rectangular tilesizes, Please change to `Array<Number>` with `[256, 256]`

File: src/util/Caching.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    statics: {
+        /**
+         * Call this function from the console (in one of your applications)
+         * and it returns config information that may be used to quickly setup a
+         * matching GWC gridset in the GeoServer.
+         *
+         * Note: If the extent is not passed to this function, the ol.view of
+         * the ol.map should have explicitly set an 'extent' property!
+         *
+         * @param {Number} tileSize The tile size in pixels
```
Comment (marcjansen) [inline src/util/Caching.js]: One does not need to have rectangular tilesizes, Please change to `Array<Number>` with `[256, 256]`

File: src/util/Caching.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+         * @param {Array} extent Optional: The extent to use for the gridSet
```
Comment (marcjansen) [inline src/util/Caching.js]: Can you hint the order of the numbers: [b, l, r, t]? So it is clear what one has to give?

File: src/util/Caching.js (hunk 1 above)
Comment (marcjansen) [reply src/util/Caching.js]: I think the docs are OK they are right now.

**Comment by marcjansen** [comment]
Hey @buehner, I have made some changes to this, and would ask you to test them in your project, if possible. The test data you send me, passes and the API of the method you needed most has not changed.

Can you please take a look?

If you approve the current state, feel free to merge this.

**Comment by marcjansen** [comment]
Wait a second, I want to get rid of the package-lock.json…

**Comment by marcjansen** [comment]
Ok, you can go ahead with a review, @buehner. TIA!

File: src/util/Caching.js [hunk 6]
```diff
@@ -82,48 +81,193 @@ Ext.define('BasiGX.util.Caching', {
...
+                    ' defaulting to [256, 256] now');
+                tileSize = [256, 256];
+            }
+            return tileSize;
+        },
 
-            return [minX, tileOriginY];
+        /**
+         * Sanitizes the passed extent, potentially taking into account the
+         * passed `ol.View`. A valid extent (left, bottom, right, top) will
```
Comment (buehner) [inline src/util/Caching.js]: `(left, bottom, right, top)` is a bit confusing. something like `(minX, minY, maxX, maxY)` would be more clear.

**Comment by buehner** [comment]
@marcjansen I found a bug and fixed it in 19e7890. I also added a test for this scenario. Please review my change and merge if you are happy. Thx

**Comment by marcjansen** [comment]
Thanks @buehner

**Pull request merged**